walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,117 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+  private ArrayList<TimeRange> getMergedTimes(ArrayList<TimeRange> takenTimes) {
+    Collections.sort(takenTimes, TimeRange.ORDER_BY_START);
+    ArrayList<TimeRange> takenTimesMerged = new ArrayList();
+    if (takenTimes.size() != 0) {
+      TimeRange previous = takenTimes.get(0);
+
+      for (TimeRange time : takenTimes) {
+        if (time.overlaps(previous)) {
+          int start = previous.start();
+          int end = Math.max(time.end(), previous.end());
+          previous = TimeRange.fromStartEnd(start, end, false);
+        } else {
+          takenTimesMerged.add(previous);
+          previous = time;
+        }
+      }
+      takenTimesMerged.add(previous);
+    }
+    return takenTimesMerged;
+  }
+
+  private ArrayList<TimeRange> getBetweenTimes(ArrayList<TimeRange> takenTimesMerged) {
+    int END_OF_DAY = 1440;
+    int START_OF_DAY = 0;
+    TimeRange emptyAtStart = TimeRange.fromStartDuration(START_OF_DAY,0);
+    TimeRange emptyAtEnd = TimeRange.fromStartDuration(END_OF_DAY, 0);
+    takenTimesMerged.add(0, emptyAtStart);
+    takenTimesMerged.add(emptyAtEnd);
+    
+    ArrayList<TimeRange> validTimes = new ArrayList();
+    for (int i = 0; i < takenTimesMerged.size() - 1; i++) {
```

Comment:
Does this run without erroring? I would expect it to error at `TimeRange adjRange = takenTimesMerged.get(i + 1);` since `i + 1` won't always be a valid index. To fix this you just need to change this for loop to stop at `takenTimesMerged.size() - 2`.

Reply:
This runs without error surprisingly

Reply:
Changing to it to -1 actually causes the last element to be ignored.

Reply:
Changing it to -2 you mean? And that should be fine since the last element should be `emptyAtEnd`. Still, that's pretty surprising that it runs without errors...

Reply:
Ooooh, you have `<`, not `<=`. Nvm then this looks good. My bad > . <